src/lifecycle.ts

```diff
@@ -13,16 +13,22 @@ export type Lifecycles =
   'beforeUpdate' | 'updated' |
   'activated' | 'deactivated'
 
-export type RouterLifecyle =
-  'beforeRouteEnter' |
-  'beforeRouteLeave' |
-  'beforeRouteUpdate'
+export type BeforeRouteUpdateLifecycle = 'beforeRouteUpdate'
+export type BeforeRouteLeaveLifecycle = 'beforeRouteLeave'
+export type BeforeRouteEnterLifecycle = 'beforeRouteEnter'
 
-export type NextFunc = ((vm: Vue) => void) | (() => void)
-export type RouterHandler = (to: any, from: any, next: NextFunc) => void
+export declare type NextFuncVmCallback<T extends Vue> = (vm: T) => void
```

Comment:
Can we merge these `NextFuncXX` into one function? I believe there is some case which variance check requirement doesn't meet but can you give me some examples?

Reply:
I want to have a stronger type check on 3 different types of lifecycles, examples can be found in the changes in README.md file, or like below:

```ts
// for `beforeRouteEnter`:
next()
// or
next(() => { ... })
// or
next((vm) => { vm.xxx() })
```

```ts
// for `beforeRouteUpdate`, only:
next()
```

```ts
// for `beforeRouteLeave`:
next()
// or
next(false)
```

Reply:
Why not 

```ts
type Next<V> = (ok?: boolean | Callback<V>) => void?
```

Reply:
Because my major point is to strictly type-check by the lifecycle. e.g. Throw error when using `next((vm) => ...)` in a `"beforeRouteUpdate"` lifecycle.